README.md
# Zombie Survivor

version: 0.2

## 모드 설명

* 하루가 지날 때마다 더 많은 좀비가 나타납니다!
* 10일마다 일부 좀비가 더 똑똑해집니다.
* 좀비는 점점 강해집니다.

## 설정 파일

* **Spawn Count Type**: 좌표 별 좀비 스폰 시도 횟수 설정 방법
    * `DEFAULT`: 3 * spawn increase amount * 일수
    * `RANDOM`: rand(0 ~ spawn increase amount * 일수)
    * 기본값: DEFAULT


* **Spawn Increase Amount**: 날이 지날 때, 증가되는 좀비 스폰 량
    * 범위: 0 ~ 10000
    * 기본값: 50


* **Random Coord Type**: 각 기준 좌표에서 랜덤 좌표를 설정 방법. 기준 좌표의 x, z 좌표에 해당 수식으로 나온 값을 더함.
    * `TRIANGLE`: rand(0 ~ 5) - rand(0 ~ 5): -5 ~ 5 중에서 0이 나올 확률이 높음. 바닐라 마크 방식
    * `UNIFORM`: rand(-5 ~ 5): -5 ~ 5가 골고루 나옴.
    * 기본값: TRIANGLE


* **??? Spawn Weight**: 해당 몬스터의 스폰 비율 가중치를 설정합니다.


* **Zombie Following Range Adder**: 강화할 좀비 인식 사거리
    * 기본값: 200.0
    * 버그: 청크 로딩 때문에 잘 작동 안함.
* **Zombie Add Following Range Day**: 인식 사거리를 강화하기 시작하는 일 수
    * 기본값: 0


* **Zombie Speed Adder**: 강화할 좀비 속도
    * 만약 Zombie Speed Up Per Day가 5이고, Zombie Speed Adder가 0.3이라면 5일마다 기본 속도의 0.3, 0.6, 0.9, ... 배가 더해짐
    * 즉, 15일 째라면 `좀비 속도 = 좀비 기본 속도 * (1 + 0.9)`
    * 기본값: 0.3
* **Zombie Speed Up Per Day**: 해당 일 수 마다 좀비 속도 강화
    * 기본값: 5
* **Zombie Max Speed Adder**: 좀비의 기본 속도에서 최대 몇배까지 더할지 설정
    * 기본값: 10.0
* **Zombie Set Or Break Block Day**: 좀비가 블록을 부수거나 설치하기 시작하는 일수
    * 기본값: 10

changelog.txt
# Zombie Survivor mc1.19.3-forge Changelog

0.1
====
 - 0.1: 하루가 지날 때마다 더 많은 좀비가 나타납니다.
        -- 서바이벌 상태의 각 플레이어 주위에 기준 좌표를 세웁니다.
        -- 각 기준 좌표의 y 좌표가 주변 좌표에 비해 낮다면 해당 기준 좌표는 좀비를 스폰하지 않습니다.
        -- 통과된 각 기준 좌표 별로 근처 랜점 좌표를 여러개 설정합니다,
        -- 랜덤 좌표 별로 좀비 소환 조건이 맞으면 (바이옴, 밝기, 플레이어와의 거리 등..) 좀비 스폰을 시도합니다.
        -- 기준 좌표 별 랜덤 좌표 갯수는 날마다 30씩 증가합니다.

0.2
====
 - 0.2: 좀비 스폰 밸런스 관련 콘피그 파일을 추가했습니다.
        -- Spawn Count Type: 좌표 별 좀비 스폰 시도 횟수 설정 방법 (DEFAULT, RANDOM) 기본값: DEFAULT
        -- Spawn Increase Amount: 날이 지날 때, 증가되는 좀비 스폰 량 (0 ~ 10000) 기본값: 10
        -- Random Coord Type: 각 기준 좌표에서 랜덤 좌표를 설정 방법 (TRIANGLE, UNIFORM) 기본값: TRIANGLE
        이제 가끔씩 주민 좀비가 스폰됩니다.
 - 0.2.1: 모드 업데이트 체크 기능 추가
 - 0.2.2: 플레이어 조건을 관전자 모드만 제외로 변경
          좀비 스폰량 관련 버그 수정 및 Spawn Increase Amount 기본값을 50으로 변경
          이제 다른 몬스터도 스폰됨. 콘피그 파일에서 스폰 비율 설정 가능

0.3
====
 - 0.3.0: 좀비 인식 사거리 강화
 - 0.3.1: 몬스터 스폰 알고리즘 약간 수정
          시간이 지날 때마다 좀비 속도 증가
 - 0.3.2: (기본값) 10일 후, 좀비가 블럭을 부수거나 설치하기 시작함.
          몬스터 스폰 알고리즘 개선
          spawnMonster, daytime 커맨드 추가
          월드 재접속 시, 좀비를 강화한 부분이 초기화되는 버그 수정


src/main/java/com/reasure/zomsurvival/util/SpawnConfig.java
package com.reasure.zomsurvival.util;

import net.minecraftforge.common.ForgeConfigSpec;


public class SpawnConfig {
    public static final ForgeConfigSpec.Builder BUILDER = new ForgeConfigSpec.Builder();
    public static final ForgeConfigSpec SPEC;

    public static final ForgeConfigSpec.EnumValue<ModEnums.SpawnCountType> SPAWN_COUNT_TYPE;
    public static final ForgeConfigSpec.IntValue SPAWN_INCREASE_AMOUNT;
    public static final ForgeConfigSpec.EnumValue<ModEnums.RandomCoordType> RANDOM_COORD_TYPE;

    public static final ForgeConfigSpec.IntValue ZOMBIE_SPAWN_WEIGHT;
    public static final ForgeConfigSpec.IntValue ZOMBIE_VILLAGER_SPAWN_WEIGHT;
    public static final ForgeConfigSpec.IntValue HUSK_SPAWN_WEIGHT;
    public static final ForgeConfigSpec.IntValue DROWNED_SPAWN_WEIGHT;
    public static final ForgeConfigSpec.IntValue SKELETON_SPAWN_WEIGHT;
    public static final ForgeConfigSpec.IntValue STRAY_SPAWN_WEIGHT;
    public static final ForgeConfigSpec.IntValue CREEPER_SPAWN_WEIGHT;
    public static final ForgeConfigSpec.IntValue SPIDER_SPAWN_WEIGHT;
    public static final ForgeConfigSpec.IntValue WITCH_SPAWN_WEIGHT;

    public static final ForgeConfigSpec.DoubleValue ZOMBIE_FOLLOWING_RANGE_MODIFIER;
    public static final ForgeConfigSpec.IntValue ZOMBIE_ADD_FOLLOWING_RANGE_DAY;

    public static final ForgeConfigSpec.DoubleValue ZOMBIE_SPEED_ADDER;
    public static final ForgeConfigSpec.IntValue ZOMBIE_SPEED_UP_PER_DAY;

    public static final ForgeConfigSpec.DoubleValue ZOMBIE_MAX_SPEED_ADDER;

    public static final ForgeConfigSpec.IntValue ZOMBIE_SET_OR_BREAK_BLOCLK_DAY;

    static {
        BUILDER.push("Configs for Zombie Survivor mod");

        SPAWN_COUNT_TYPE = BUILDER.comment("기준 좌표 별 좀비 스폰 시도 횟수 설정 방법")
                .comment("DEFAULT: 3 * spawn increase amount * 일수")
                .comment("RANDOM: rand(0 ~ spawn increase amount * 일수)")
                .defineEnum("Spawn Count Type", ModEnums.SpawnCountType.DEFAULT);

        SPAWN_INCREASE_AMOUNT = BUILDER.comment("날이 지날 때, 증가되는 좀비 스폰 량")
                .defineInRange("Spawn Increase Amount", 50, 0, 10000);

        RANDOM_COORD_TYPE = BUILDER.comment("각 기준 좌표에서 랜덤 좌표를 설정 방법. 기준 좌표의 x, z 좌표에 해당 수식으로 나온 값을 더함.")
                .comment("TRIANGLE: rand(0 ~ 5) - rand(0 ~ 5): -5 ~ 5 중에서 0이 나올 확률이 높음. 마크 바닐라 방식")
                .comment("UNIFORM: rand(-5 ~ 5): -5 ~ 5가 골고루 나옴.")
                .defineEnum("Random Coord Type", ModEnums.RandomCoordType.TRIANGLE);

        ZOMBIE_SPAWN_WEIGHT = BUILDER.comment("좀비 스폰 가중치")
                .defineInRange("Zombie Spawn Weight", 1000, 0, Integer.MAX_VALUE);

        ZOMBIE_VILLAGER_SPAWN_WEIGHT = BUILDER.comment("주민 좀비 스폰 가중치")
                .defineInRange("Zombie Villager Spawn Weight", 100, 0, Integer.MAX_VALUE);

        HUSK_SPAWN_WEIGHT = BUILDER.comment("허스크 스폰 가중치")
                .defineInRange("Husk Spawn Weight", 250, 0, Integer.MAX_VALUE);

        DROWNED_SPAWN_WEIGHT = BUILDER.comment("드라운드 스폰 가중치")
                .defineInRange("Drowned Spawn Weidht", 350, 0, Integer.MAX_VALUE);

        SKELETON_SPAWN_WEIGHT = BUILDER.comment("스켈레톤 스폰 가중치")
                .defineInRange("Skeleton Spawn Weight", 0, 0, Integer.MAX_VALUE);

        STRAY_SPAWN_WEIGHT = BUILDER.comment("스트레이 스폰 가중치")
                .defineInRange("Stray Spawn Weight", 0, 0, Integer.MAX_VALUE);

        CREEPER_SPAWN_WEIGHT = BUILDER.comment("크리퍼 스폰 가중치")
                .defineInRange("Creeper Spawn Weight", 0, 0, Integer.MAX_VALUE);

        SPIDER_SPAWN_WEIGHT = BUILDER.comment("거미 스폰 가중치")
                .defineInRange("Spider Spawn Weight", 0, 0, Integer.MAX_VALUE);

        WITCH_SPAWN_WEIGHT = BUILDER.comment("마녀 스폰 가중치")
                .defineInRange("Witch Spawn Weight", 0, 0, Integer.MAX_VALUE);

        ZOMBIE_FOLLOWING_RANGE_MODIFIER = BUILDER.comment("강화할 좀비 인식 사거리")
                .defineInRange("Zombie Following Range Adder", 200.0, 0.0, Double.MAX_VALUE);

        ZOMBIE_ADD_FOLLOWING_RANGE_DAY = BUILDER.comment("좀비 인식 사거리 강화를 시작할 일 수")
                .defineInRange("Zombie Add Following Range Day", 0, 0, Integer.MAX_VALUE);

        ZOMBIE_SPEED_ADDER = BUILDER.comment("Zombie Speed Up Per Day 마다 강화될 속도 (바닐라에서 좀비의 기본 속도는 0.23)")
                .comment("Ex) Zombie Speed Up Per Day가 5일 이고, Zombie Speed Adder가 0.1이고, 현재 15일 일경우, 0.23 + 0.23 * 0.1배 * 3일")
                .comment("즉, 기존 속도의 0.1배, 0.2배, 0.3배, ... 가 더해지는 식으로 강화됨.")
                .defineInRange("Zombie Speed Adder", 0.3D, 0.0D, 10.0D);

        ZOMBIE_SPEED_UP_PER_DAY = BUILDER.comment("해당 일 수마다 좀비 속도가 강화됨.")
                .defineInRange("Zombie Speed Up Per Day", 5, 0, Integer.MAX_VALUE);

        ZOMBIE_MAX_SPEED_ADDER = BUILDER.comment("좀비의 기본 속도에서 최대 몇배까지 더할지 설정")
                .defineInRange("Zombie Max Speed Adder", 10.0, 0.0, 100.0);

        ZOMBIE_SET_OR_BREAK_BLOCLK_DAY = BUILDER.comment("좀비가 블록을 부수거나 설치하기 시작하는 일 수")
                .defineInRange("Zombie Set Or Break Block Day", 10, 0, Integer.MAX_VALUE);

        BUILDER.pop();
        SPEC = BUILDER.build();
    }
}
